### PR TITLE
replace all ${VAR} references with env var values in watches yaml

### PR DESCRIPTION
**Pull request by jmazzitelli**
**Description of the change:**

For any ${VAR} references in the watches yaml, this replaces those env vars with their values before the watches yaml is processed. If there is no ${VAR} defined, the "${VAR}" remains as-is in the watches yaml.

**Motivation for the change:**

There are times when a user wants to customize what an operator will watch. Right now, watches.yaml is fixed and cannot be customized or configured. This feature at least allows the operator to be somewhat configurable by passing in env vars in the operator's subscription or directly in its Deployment.

For one such example, I may have an operator that can be deployed on OpenShift and non-OpenShift environments. I therefore could have two different sets of playbooks in my operator container - one for OpenShift and one for non-OpenShift. I could define a set of ${WATCH_PLAYBOOK_XYZ} env vars and point to cluster-specific playbooks in the watches.yaml:
```yaml
- version: v1
  group: bar.example.com
  kind: Bar
  playbook: ${WATCH_PLAYBOOK_BAR}
```
When deployed on OpenShift, my OLM CSV could define that env var to "playbook-bar-openshift.yml" and on non-OpenShift, I have its OLM CSV define the env var to "playbook-bar-kubernetes.yml".

Another example. I want my operator to watch Bar objects, but I do NOT want the operator to run my playbook reconciliation on every Bar in the cluster - I only want to run a reconciliation when a Bar with a specific selector. But! (and here's the important part), this selector can be different depending on what the admin wants (the admin being the person who installed the operator). With this PR, this can be accomplished like this:

[watches.yaml](https://sdk.operatorframework.io/docs/building-operators/ansible/reference/watches/):
```yaml
- version: v1
  group: bar.example.com
  kind: Bar
  playbook: playbook.yml
  selector:
    matchLabels:
      ${SPECIAL_BAR_NAME}: ${SPECIAL_BAR_VAL}
```

Operator Subscription:

```yaml
kind: Subscription
metadata:
  name: my-operator
spec:
  package: my-operator
  channel: stable
  config:
    env:
    - name: SPECIAL_BAR_NAME
      value: "special-bar-name"
    - name: SPECIAL_BAR_VAL
      value: "special-bar-value"
```

### CODE REVIEW

File: internal/ansible/watches/watches.go [hunk 1]
```diff
@@ -358,6 +362,32 @@ func Load(path string, maxReconciler, ansibleVerbosity int) ([]Watch, error) {
 	return watches, nil
 }
 
+// replaceEnvVariables will replace all ${VAR} references found in the byte array
+// with the actual values of the env variables. If an env variable is not defined
+// the ${VAR} reference remains as-is in the returned byte array.
+func replaceEnvVariables(input []byte) []byte {
+	// Regular expression to match "${VARIABLE}"
+	envVarPattern := regexp.MustCompile(`\${([^}]+)}`)
```
Comment (everettraven) [inline internal/ansible/watches/watches.go:370]: I'm a bit concerned about using/maintaining regex here. I wonder if there is a good way for us to use go templates in the yaml as opposed to finding regex matches for "${...}".

I haven't had much time to dig into this thought more, but I can play around with some stuff tomorrow and see if we can get something a bit more robust

Comment (jmazzitelli) [reply internal/ansible/watches/watches.go:370]: What's the concern over using a regex pattern here? It's a fairly innocuous regex pattern that is just picking out the `${var}` syntax. It's a pretty typical regex pattern that would likely never have to change (certainly that regex pattern is not anything out of the ordinary).

Comment (jmazzitelli) [reply internal/ansible/watches/watches.go]: @everettraven check out the latest commit - removes regex dependency. Simply uses os.Expand functionality instead.

Comment (everettraven) [reply internal/ansible/watches/watches.go]: > What's the concern over using a regex pattern here? It's fairly innocuous that is just picking out the `${var}` syntax. It's a pretty typical pattern that would likely never have to change (certainly that regex pattern is not anything out of the ordinary).

My concern is simply maintaining the regex. In my experience using custom regex over other methods can result in unexpected behavior and bugs. For example, this regex currently matches all the following:
- ${SOMETHING}
- ${SOMETHING_COOL}
- ${NICE-REG}
- ${{{{Some cool thing}}
- ${5Things}
- ${TheThing=AnotherThing}

I'm by no means an expert in environment variable naming and what is allowed and what is not but some of these seem like they would be invalid (please do correct me if I'm wrong!). For every one of these edge cases the regex syntax gets more and more complex.

Using the os.Expand functionality is an approach I'm personally more comfortable with, especially because it is part of the standard library.

File: internal/ansible/watches/watches.go [hunk 2]
```diff
@@ -358,6 +361,22 @@ func Load(path string, maxReconciler, ansibleVerbosity int) ([]Watch, error) {
...
 
+// replaceEnvVariables will replace all ${VAR} references found in the byte array
+// with the actual values of the env variables. If an env variable is not defined
+// the ${VAR} reference remains as-is in the returned byte array.
+func replaceEnvVariables(input []byte) []byte {
+	outputString := os.Expand(string(input), func(varName string) string {
+		varValue, exists := os.LookupEnv(varName)
+		if !exists {
+			// If the environment variable doesn't exist, keep the original reference
+			return "${" + varName + "}"
```
Comment (everettraven) [inline internal/ansible/watches/watches.go:372]: Is this a typical pattern when doing environment variable substitution? I would have expected if it doesn't exist that an empty value is used.

Comment (jmazzitelli) [reply internal/ansible/watches/watches.go:372]: Yes, that is actually the [default behavior of Go](https://pkg.go.dev/os#ExpandEnv) if I were to use function [os.ExpandEnv()](https://cs.opensource.google/go/go/+/go1.21.5:src/os/env.go;l=50). I can very easily change this code so it uses that. Let me know if you want that behavior instead.

Let me describe what my thought process was here, and why the code keeps the env var reference as-is rather than replace it with empty string. I was looking at this more for backward compatibility. Think about how this would work today. If someone has the string `${FOO}` already in their watches today (why would they do that? I have no idea :) but suppose they did) then today the operator would see that literal value of `${FOO}` (because currently the operator does nothing with watches.yaml other than read it all in and use the yaml text as-is).

This PR's new behavior wouldn't change this. Unless the user actually sets FOO env var on the operator (and presumably they do not - why would they when watches.yaml does not currently substitute these anyway), this new behavior would mimic the same behavior as today (that is, the operator would see `${FOO}` in the watches.yaml).

But again, I'm OK with changing this and going to the default Go behavior which is - set it to an empty string if there is no FOO env var.

Comment (everettraven) [reply internal/ansible/watches/watches.go:372]: Gotcha. I think the backwards compatibility reasoning is sufficient, although like you said I'm not sure why someone would have done that but 🤷 .

File: internal/ansible/watches/watches_test.go [hunk 3]
```diff
@@ -892,3 +892,49 @@ func TestGetPossibleRolePaths(t *testing.T) {
...
+	os.Setenv("WATCH_PLAYBOOK", playbookFile)
+	defer os.Unsetenv("WATCH_PLAYBOOK")
+	os.Setenv("WATCH_VERSION", "v123")
+	defer os.Unsetenv("WATCH_VERSION")
+	os.Setenv("WATCH_MATCH_LABEL_VAR_NAME", "label123")
+	defer os.Unsetenv("WATCH_MATCH_LABEL_VAR_NAME")
+	os.Setenv("WATCH_MATCH_LABEL_VAR_VALUE", "value123")
+	defer os.Unsetenv("WATCH_MATCH_LABEL_VAR_VALUE")
+	os.Setenv("WATCH_MATCH_EXPRESSIONS_KEY", "key123")
+	defer os.Unsetenv("WATCH_MATCH_EXPRESSIONS_KEY")
```
Comment (everettraven) [inline internal/ansible/watches/watches_test.go:913]: I'd use [`t.Setenv(...)`](https://pkg.go.dev/testing#T.Setenv) here. It will handle the cleanup and reset these to previous values if for some reason they already exist on someone's machine.

File: internal/ansible/watches/watches.go (hunk 2 above)
Comment (jmazzitelli) [reply internal/ansible/watches/watches.go:372]: The only issue I see with this way of using os.Expand versus using the regex is that (again looking at it for backward compatibility) if the user already had `$FOO` in the watches.yaml (notice: NO braces - it is NOT `${FOO}` but rather just `$FOO`), and there is no FOO env var set in the operator, then this will get changed to `${FOO}`.

This is because os.Expand will look for env var references in both formats (`$FOO` and `${FOO}`) but it doesn't have a way to tell you which form is found. It just passes the env var name to the function. So the code has no way of knowing if it was specified with or without braces. So this PR will simply assume braces and that's why it does `"${" + varName + "}"`

So, in short, if someone, for whatever reason, already has `$FOO` in their watches.yaml and does NOT set FOO env var in the operator, the watches.yaml will have it replaced with `${FOO}` - so not 100% backward compat.

File: internal/ansible/watches/watches_test.go (hunk 3 above)
Comment (jmazzitelli) [reply internal/ansible/watches/watches_test.go:913]: k.. I can do that. I was just trying to stay consistent with some of the test code I already saw in here. e.g.

line 486ff:
```
os.Setenv("WORKER_MAXCONCURRENTRECONCILESENV_APP_EXAMPLE_COM", "4")
defer os.Unsetenv("WORKER_MAXCONCURRENTRECONCILESENV_APP_EXAMPLE_COM")
os.Setenv("ANSIBLE_VERBOSITY_ANSIBLEVERBOSITYENV_APP_EXAMPLE_COM", "4")
defer os.Unsetenv("ANSIBLE_VERBOSITY_ANSIBLEVERBOSITYENV_APP_EXAMPLE_COM")
```

However, using t.Setenv is a nicer way to do it.

File: internal/ansible/watches/watches_test.go [hunk 4]
```diff
@@ -892,3 +892,49 @@ func TestGetPossibleRolePaths(t *testing.T) {
 		})
 	}
 }
+
+func TestReplaceEnvVars(t *testing.T) { //nolint:gocyclo
```
Comment (everettraven) [inline internal/ansible/watches/watches_test.go:896]: Could you elaborate on the reasoning for the nolint here?

**Review by everettraven — COMMENTED** [review]
Overall looks good to me.  Couple comments:

File: internal/ansible/watches/watches_test.go (hunk 3 above)
Comment (jmazzitelli) [reply internal/ansible/watches/watches_test.go]: this is done.

File: internal/ansible/watches/watches_test.go (hunk 4 above)
Comment (jmazzitelli) [reply internal/ansible/watches/watches_test.go:896]: Copy-paste :) I copied the TestLoad function when I started writing my test function, and [it has that](https://github.com/operator-framework/ansible-operator-plugins/blob/main/internal/ansible/watches/watches_test.go#L108C1-L109C1). No reason other than that. I can delete that.

Comment (jmazzitelli) [reply internal/ansible/watches/watches_test.go]: this is done in latest commit

**Review by everettraven — APPROVED** [review]

**Comment by jmazzitelli** [comment]
Addressed all the suggestions.

As for release notes, I would say we need to mention that odd edge case where `$FOO` (without braces) will be replaced with `${FOO}` (with braces) if FOO env var is not set in the operator. So the release notes should say something like:

```
Environment variable references of the form `$ENV` or `${ENV}` found in
watches.yaml will be replaced with the values of those environment variables
if they are set in the operator. If the environment variable is not set in the operator,
the watches.yaml will keep the string `${ENV}` as-is. Take note that if your
watches.yaml has a string in the form `$FOO` (without braces) and you do
not have an environment variable named FOO set in the operator, then the
watches.yaml will effectively replace `$FOO` with the string `${FOO}` (i.e. with
braces surrounding the name).
```

**Comment by everettraven** [comment]
We currently have autogenerated release notes in this repository, but once the changes are pulled into the Operator-SDK we can provide a more detailed release note entry. Thanks for the contribution @jmazzitelli !
